### DISCUSSION TITLE
ListView Not Virtualizing

### DISCUSSION THREAD

**Discussion by hamu777** [microsoft/microsoft-ui-xaml#10138] [Q&A] [answered]
I have enabled VirtualizingStackPanel in my WinUI 3 ListView, but it seems that virtualization is not working. In the InitializeComponent function of the window containing the ListView, I checked the virtualization status using the following code:

```C++
bool IsVirtualizing = winrt::Microsoft::UI::Xaml::Controls::VirtualizingStackPanel::GetIsVirtualizing( MyListView() );
```
However, it returns false, indicating that virtualization is not enabled. Here is the XAML code for the ListView:

```XML
    <Grid>
        <ListView x:Name="MyListView">
            <ListView.ItemsPanel>
                <ItemsPanelTemplate>
                    <VirtualizingStackPanel />
                </ItemsPanelTemplate>
            </ListView.ItemsPanel>
            <ListView.ItemTemplate>
                <DataTemplate>
                    <TextBlock Text="{Binding}"/>
                </DataTemplate>
            </ListView.ItemTemplate>
        </ListView>
    </Grid>
```

Could you please help me identify what might be wrong or what I might be missing to enable virtualization?
I am trying to display 10,000 data entries. 

```C++
void MainWindow::InitializeComponent( void )
{
  MainWindowT<MainWindow>::InitializeComponent();

  auto items = winrt::single_threaded_observable_vector<winrt::hstring>();

  for( int i = 0; i < 10000; i++ )
  {
    items.Append( hstring{ std::to_wstring( i ) } );
  }

  MyListView().ItemsSource(items);

  bool IsVrtualizing = winrt::Microsoft::UI::Xaml::Controls::VirtualizingStackPanel::GetIsVirtualizing( MyListView() );
  if (IsVrtualizing)
  {
    OutputDebugString(L"\n IsVrtualizing = True\n");
  }
  else
  {
    OutputDebugString(L"\n IsVrtualizing = False\n");
  }
}
```

Thank you!

**Comment by Petrarca181** [comment]
Stupid queestion, isn't listview virtualizing items by default in wiunui 3? What is VirtualizingStackPanel for? Just for curiosity.

**Reply by hamu777** [reply]
I initially assumed that items were virtualized by default, but when I checked with GetIsVirtualizing, it returned false. From the documentation, I understand that virtualization is enabled by default in WPF, but I couldn’t find any documentation confirming this for WinUI 3. Therefore, I’m currently testing virtualization manually, as it seems that it might not be enabled by default.

**Reply by Petrarca181** [reply]
Default ListView is using  ItemsStackPanel, it should be virtualized. 
```
<Setter Property="ItemsPanel">
            <Setter.Value>
                <ItemsPanelTemplate>
                    <ItemsStackPanel Orientation="Vertical" />
                </ItemsPanelTemplate>
            </Setter.Value>
        </Setter>
```

Idk how to check it, but In my app I have big lists and don't see any performance drop. Isn't your list in some sort of parent that allows to render the entire list without  limiting it's height?

**Reply by hamu777** [reply]
I would be pleased if it is indeed virtualized as expected. However, The entire list is contained within a Grid, which should restrict its height. Although the performance appears fine, I couldn’t locate any documentation on this topic and remain concerned as to whether virtualization is truly active.
However, since there are no problems with performance at the moment, it may be okay to think of it as virtualized for the time being, but I wanted to ask here just to be sure.
As mentioned in my original question, should I be worried that GetIsVirtualizing returns False, or is this unrelated to the virtualization behavior of the ListView?

**Reply by hamu777** [reply, marked as answer]
It was stated on the ItemsStackPanel description page that virtualization is enabled, which has put my mind at ease. Thank you for listening to my concerns.
